shims/spark32/src/main/scala/org/apache/spark/sql/execution/adaptive/GlutenCostEvaluator.scala

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.joins.ShuffledJoin
+
+/**
+ * This [[CostEvaluator]] is to force use the new physical plan when AQE's new physical plan
+ * contains [[ShuffledJoin]] and cost is the same.
+ */
+case class GlutenCostEvaluator() extends CostEvaluator {
+  private var isReOptimize = false
+  override def evaluateCost(plan: SparkPlan): Cost = {
+    val simpleCost =
+      SimpleCostEvaluator.evaluateCost(plan).asInstanceOf[SimpleCost]
+    var cost = simpleCost.value * 2
+    if (isReOptimize && existsShuffledJoin(plan)) {
+      cost -= 1
+    }
+    isReOptimize = !isReOptimize
```

Comment:
I could understand how this statement works but it's like something too caller-sensitive to rely on.

I see `Cost` is a trait that can be extended. Could we store the plan's id which is incremental by instantiation order in a customized Cost structure, then use the ID during cost comparison to find out the one that was created later?

Reply:
The planId is increasing, but it may be removed by the parent node, causing the id to become smaller, such as `OptimizeOneRowPlan`,
I'm going to do it this way, what do you think?
```
case class GlutenCost(value: Long, planId: Int) extends Cost {
  override def compare(that: Cost): Int = that match {
    case GlutenCost(thatValue, thatId) =>
      if (value < thatValue || (value == thatValue && planId != thatId)) -1
      else if (value > thatValue) 1
      else 0
    case _ =>
      throw QueryExecutionErrors.cannotCompareCostWithTargetCostError(that.toString)
  }
}
```